torch/distributed/rendezvous.py

```diff
@@ -88,6 +93,33 @@ def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
         raise RuntimeError("No rendezvous handler for {}://".format(result.scheme))
     return _rendezvous_handlers[result.scheme](url, **kwargs)
 
+def _create_store_from_options(backend_options, rank):
+    result = urlparse(backend_options.init_method)
+
+    # If using env initialization, get rank and world_size from env
+    world_size = -1
+    if result.scheme == "env":
+        rank = os.environ.get("RANK", rank)
+        # world_size default set to -1
+        # if the world_size env variable is not present then it is a dynamic group
+        world_size = int(os.environ.get("WORLD_SIZE", world_size))
+
+    query_dict = _query_to_dict(result.query)
+    if rank != -1:
+        query_dict["rank"] = rank
```

Comment:
Could you please elaborate the logic here? Would I be correct that we currently require applications to specify a rank regardless of whether this is dynamic grouping or not?

Reply:
BTW, would I be correct that we need a TODO item to automatically assign ranks? Because with dynamic world size, a developer might not know the rank for the process?

Reply:
> Could you please elaborate the logic here? Would I be correct that we currently require applications to specify a rank regardless of whether this is dynamic grouping or not?

Yes rank is always required. However, it does not necessarily need to be passed into `init_rpc` as `RANK` and `WORLD_SIZE` can be set in environment variables and used during `env://` initialization. Therefore we run into the issue of needing to check that environment variables for the rank value and setting it to that. If rank is not set in either `init_rpc` or `env` then it will be missing from the reformatted query and will throw an error in `_rendezvous_handlers` with error `rank parameter is missing`

Reply:
> BTW, would I be correct that we need a TODO item to automatically assign ranks? Because with dynamic world size, a developer might not know the rank for the process?

Yes, an additional feature would be to assign ranks internally so users do not need to assign the ranks themselves.